File: go/utils/grid_calculation.go
package utils

import (
	"fmt"
	
	"github.com/sam/fm-data-capture/models"
)

// GridItem represents a position in the grid
type GridItem struct {
	Row int
	Col int
}

// CalculateGridItem calculates which grid item was clicked based on coordinates
func CalculateGridItem(x, y int) *GridItem {
	gridStartX := 2072
	gridStartY := 620
	boxWidth := 73
	boxHeight := 73
	paddingX := 24
	paddingY := 25
	numRows := 13
	numCols := 3
	
	// Check if click is within grid boundaries
	if !(gridStartX <= x && x <= gridStartX+(boxWidth+paddingX)*numCols-paddingX) {
		if x < gridStartX {
			fmt.Printf("Click x-coordinate %d is left of the grid boundaries.\n", x)
		} else {
			fmt.Printf("Click x-coordinate %d is right of the grid boundaries.\n", x)
		}
		return nil
	}
	
	if !(gridStartY <= y && y <= gridStartY+(boxHeight+paddingY)*numRows-paddingY) {
		if y < gridStartY {
			fmt.Printf("Click y-coordinate %d is above the grid boundaries.\n", y)
		} else {
			fmt.Printf("Click y-coordinate %d is below the grid boundaries.\n", y)
		}
		return nil
	}
	
	// Calculate the row and column
	col := (x - gridStartX) / (boxWidth + paddingX)
	row := (y - gridStartY) / (boxHeight + paddingY)
	
	// Calculate exact coordinates of the top-left corner of the cell
	cellStartX := gridStartX + col*(boxWidth+paddingX)
	cellStartY := gridStartY + row*(boxHeight+paddingY)
	
	// Check if click is within the cell's box
	if !(cellStartX <= x && x <= cellStartX+boxWidth &&
		cellStartY <= y && y <= cellStartY+boxHeight+1) {
		fmt.Println("Clicked between grid items.")
		return nil
	}
	
	// Check if calculated row and column are within grid dimensions
	if 0 <= row && row < numRows && 0 <= col && col < numCols {
		return &GridItem{Row: row, Col: col}
	}
	
	fmt.Printf("Calculated row %d or column %d is outside the grid dimensions.\n", row, col)
	return nil
}

// MatchRuneIdByPosition matches a rune ID based on the clicked position and available characteristics
func MatchRuneIdByPosition(x, y int, caracs []models.CaracValue) int {
	gridItem := CalculateGridItem(x, y)
	if gridItem == nil {
		fmt.Println("Clicked outside the grid.")
		return 0 // Using 0 as "not found" indicator
	}
	
	if gridItem.Row < len(caracs) {
		currentCaracValue := caracs[gridItem.Row]
		for caracID, runeIDs := range models.RuneIDs {
			if currentCaracValue.ID == caracID && gridItem.Col < len(runeIDs) {
				return runeIDs[gridItem.Col]
			}
		}
		fmt.Printf("No rune found for carac %d.\n", currentCaracValue.ID)
	}
	
	return 0
}

File: grid_calculation.py
import re
from capture_functions import COMBINE_BUTTON_REGION
from rune_data import runeNumbers, caracId, runeIds


def calculate_grid_item(x, y):
    """
    Calculates the grid item clicked based on the coordinates.

    Args:
        x (int): The x-coordinate of the click.
        y (int): The y-coordinate of the click.

    Returns:
        tuple: A tuple (row, column) representing the grid item clicked, or None if the click is outside the grid.
    """
    grid_start_x = 2072
    grid_start_y = 620
    box_width = 73
    box_height = 73
    padding_x = 24  # Padding for X-axis
    padding_y = 25  # Padding for Y-axis
    num_rows = 13
    num_cols = 3

    # Check if the click is within the grid boundaries
    if not (grid_start_x <= x <= grid_start_x + (box_width + padding_x) * num_cols - padding_x):
        if x < grid_start_x:
            print(f"Click x-coordinate {x} is left of the grid boundaries.")
        else:
            print(f"Click x-coordinate {x} is right of the grid boundaries.")
        return None
    if not (grid_start_y <= y <= grid_start_y + (box_height + padding_y) * num_rows - padding_y):
        if y < grid_start_y:
            print(f"Click y-coordinate {y} is above the grid boundaries.")
        else:
            print(f"Click y-coordinate {y} is below the grid boundaries.")
        return None

    # Calculate the row and column
    col = (x - grid_start_x) // (box_width + padding_x)
    row = (y - grid_start_y) // (box_height + padding_y)

    # Calculate the exact x and y coordinates of the top-left corner of the cell
    cell_start_x = grid_start_x + col * (box_width + padding_x)
    cell_start_y = grid_start_y + row * (box_height + padding_y)

    # Check if the click is within the cell's box
    if not (cell_start_x <= x <= cell_start_x + box_width and
            cell_start_y <= y <= cell_start_y + box_height+1):
        print("Clicked between grid items.")
        return None

    # Check if the calculated row and column are within the grid dimensions
    if 0 <= row < num_rows and 0 <= col < num_cols:
        return (row, col)
    else:
        print(f"Calculated row {row} or column {col} is outside the grid dimensions.")
        return None
    



def matchRuneIdByPosition(x, y, caracs):
    """
    Matches the rune used based on the carac name at the given row.

    Args:
        row (int): The row index of the carac.
        col (int): The column index.
        caracs (list): A list of carac names.

    Returns:
        int: A rune ID if a match is found, or None if no match is found.
    """
    
    if (COMBINE_BUTTON_REGION["left"] <= x <= COMBINE_BUTTON_REGION["left"] + COMBINE_BUTTON_REGION["width"] and
            COMBINE_BUTTON_REGION["top"] <= y <= COMBINE_BUTTON_REGION["top"] + COMBINE_BUTTON_REGION["height"]):
        return 38

    grid_item = calculate_grid_item(x, y)
    if grid_item is None:
        print("Clicked outside the grid.")
        return None
    
    print(f"Clicked on grid item: {grid_item}")
    row, col = grid_item

    if row <= len(caracs):
        currentCaracValue, currentCaracId = caracs[row]
        for caracId, runeId in runeIds.items():
            if currentCaracId == caracId:
                return runeId[col]
        print(f"No rune found for carac {currentCaracId}.")

    return None

